[InteractionRegions] Generated clip paths are sometimes offset from their element

etiennesegonzac opened the pull request:
#### 2775f5bf7c3e8aeb9b44090c7a277aedcdb7d0e4
<pre>
[InteractionRegions] Generated clip paths are sometimes offset from their element
<a href="https://bugs.webkit.org/show_bug.cgi?id=273428">https://bugs.webkit.org/show_bug.cgi?id=273428</a>
&lt;<a href="https://rdar.apple.com/126865156">rdar://126865156</a>&gt;

Reviewed by Simon Fraser.

When generating Interaction Regions, the paths we build to match
shapes are complete and relative to the element&apos;s bounds.
But the Interaction Region&apos;s bounds respect clipping.
Provide the offset between the two to `interactionRegionForRenderedRegion`
so we can adjust the paths location if needed.
(Effectively moving the path in the mask layer on the UI-side.)

Also provide the current transform to skip shape generation for
rotations / skew (unsupported) and also to  avoid having to go through
`absoluteQuads` just to get a bounding size.

This applies to SVG elements, icons and CSS clip-paths.

For SVG icons (using image tags), always use the image from the
SVGImageCache to match what was just painted.

* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::boundingSize):
(WebCore::absoluteBoundingRect): Deleted.
Don&apos;t use `absoluteQuads` to get the bounding size.
(WebCore::findIconImage):
Use `imageForRenderer` to make sure icon masking is based on what was
just painted (and potentially cached in the SVGImageCache).
(WebCore::interactionRegionForRenderedRegion):
For icons, generate the raster shape at 0x0 directly instead of doing an
extra translation.
For SVG elements, use the full viewBoxTransform instead of just the
scale.
Use the new `clipOffset`.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::unite):
(WebCore::EventRegionContext::uniteInteractionRegions):
Pass an optional transform and the clip offset to
`interactionRegionForRenderedRegion`.

* LayoutTests/interaction-region/icon-masking.html:
Add test cases clipping on the top left corner (bitmap and SVG).
Add test cases with nested transforms (with and without compositing
layers).
Add a test case with rotation (no clip-path).
* LayoutTests/interaction-region/icon-masking-expected.txt:
New interaction regions, new guards on the SVG logo now matching the &quot;is
complex shape&quot; check, and floating point diff.
* LayoutTests/interaction-region/svg.html:
Add a test case clipping on the top left corner.
* LayoutTests/interaction-region/svg-expected.txt:
New interaction regions and floating point diff.
* LayoutTests/interaction-region/clip-path-expected.txt:
Add a test case clipping on the left and right edge.
Add test cases with rotation / skew (no clip-path).
* LayoutTests/interaction-region/clip-path.html:
New interaction regions.

Canonical link: <a href="https://commits.webkit.org/278371@main">https://commits.webkit.org/278371@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d006798aafa2e74576e9fdcecd01a39ac5ad0695

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/933 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/564 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52342 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27204 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22098 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24622 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/501 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8621 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46608 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55087 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25339 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48399 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43433 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47422 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11043 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27464 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -444,34 +458,34 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         clipPath = styleClipPath->getPath(TransformOperationData(FloatRect(FloatPoint(), boundingRect.size())));
     } else if (iconImage && originalElement) {
-        LayoutRect imageRect(rect);
+        auto boundingRect = absoluteBoundingRect(regionRenderer);
```
Calling `absoluteBoundingRect` just to get its size is wasteful; it pays the cost for mapping the rect to absolute coordinates.

etiennesegonzac replied:
~~we use the location in `pathTranslationToMatchClipping` below~~

no we don't, updating :)

etiennesegonzac replied:
we still need the transform but this should be better

smfr commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -441,37 +463,37 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
     if (styleClipPath && styleClipPath->type() == PathOperation::OperationType::Shape && originalElement) {
-        auto boundingRect = absoluteBoundingRect(regionRenderer);
-        clipPath = styleClipPath->getPath(TransformOperationData(FloatRect(FloatPoint(), boundingRect.size())));
+        auto size = boundingSize(regionRenderer, transform);
```
I don't know if this transform thing is correct. You've deviated so far from the painting code that it's impossible to know how much of this transform/clipping logic is OK.

smfr commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -279,6 +287,20 @@ static std::optional<std::pair<Ref<SVGSVGElement>, Ref<SVGGraphicsElement>>> fin
+    if (!clippingRect)
+        return FloatSize();
+
+    auto clipDelta = FloatSize(clippedRect.width() - boundingSize.width(), clippedRect.height() - boundingSize.height());
```
This seems really ad-hoc. What is it doing?

smfr commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -71,7 +71,15 @@ void EventRegionContext::unite(const FloatRoundedRect& roundedRect, RenderObject
+        clipRect = m_clipStack.last();
+    std::optional<AffineTransform> transform;
+    if (!m_transformStack.isEmpty())
+        transform = m_transformStack.last();
```
Is just grabbing the last transform on the stack OK? What transform do you need; a root-relative one, or one relative to some ancestor? Which ancestor?

etiennesegonzac replied:
We want to apply the same transform / clipping as the one applied to the layer bounds in `transformAndClipIfNeeded`. (it only uses the last clip rect / transform)

I'm adding a couple of tests with nested transforms

smfr commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -226,11 +226,19 @@ static bool isGuardContainer(const Element& element)
+
+    FloatSize size = unionRect(rects).size();
+    if (transform)
+        size.scale(transform->xScale(), transform->yScale());
```
Why only scale? What if the transform is a skew or rotate?

etiennesegonzac committed:
[InteractionRegions] Generated clip paths are sometimes offset from their element
https://bugs.webkit.org/show_bug.cgi?id=273428
<rdar://126865156>

Reviewed by Simon Fraser.

When generating Interaction Regions, the paths we build to match
shapes are complete and relative to the element's bounds.
But the Interaction Region's bounds respect clipping.
Provide the offset between the two to `interactionRegionForRenderedRegion`
so we can adjust the paths location if needed.
(Effectively moving the path in the mask layer on the UI-side.)

Also provide the current transform to skip shape generation for
rotations / skew (unsupported) and also to  avoid having to go through
`absoluteQuads` just to get a bounding size.

This applies to SVG elements, icons and CSS clip-paths.

For SVG icons (using image tags), always use the image from the
SVGImageCache to match what was just painted.

* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::boundingSize):
(WebCore::absoluteBoundingRect): Deleted.
Don't use `absoluteQuads` to get the bounding size.
(WebCore::findIconImage):
Use `imageForRenderer` to make sure icon masking is based on what was
just painted (and potentially cached in the SVGImageCache).
(WebCore::interactionRegionForRenderedRegion):
For icons, generate the raster shape at 0x0 directly instead of doing an
extra translation.
For SVG elements, use the full viewBoxTransform instead of just the
scale.
Use the new `clipOffset`.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::unite):
(WebCore::EventRegionContext::uniteInteractionRegions):
Pass an optional transform and the clip offset to
`interactionRegionForRenderedRegion`.

* LayoutTests/interaction-region/icon-masking.html:
Add test cases clipping on the top left corner (bitmap and SVG).
Add test cases with nested transforms (with and without compositing
layers).
Add a test case with rotation (no clip-path).
* LayoutTests/interaction-region/icon-masking-expected.txt:
New interaction regions, new guards on the SVG logo now matching the "is
complex shape" check, and floating point diff.
* LayoutTests/interaction-region/svg.html:
Add a test case clipping on the top left corner.
* LayoutTests/interaction-region/svg-expected.txt:
New interaction regions and floating point diff.
* LayoutTests/interaction-region/clip-path-expected.txt:
Add a test case clipping on the left and right edge.
Add test cases with rotation / skew (no clip-path).
* LayoutTests/interaction-region/clip-path.html:
New interaction regions.

Canonical link: https://commits.webkit.org/278371@main
